Prerender value of textarea element as text content

surayya-MS opened the pull request:
# Prerender value of textarea element as text content

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

## Description
```csharp
<textarea @bind="@value"></textarea> 
@code 
{
    string value = "Hello";
}
```

This prerenders as `<textarea value="Hello"></textarea>`,  it should be `<textarea>Hello</textarea>`.

Fixes #45967

surayya-MS committed:
prerender value attribute of textarea element as text content

SteveSandersonMS commented on src/Mvc/Mvc.ViewFeatures/src/RazorComponents/HtmlRenderer.cs:
```diff
@@ -188,6 +194,34 @@ private int RenderFrames(HtmlRenderingContext context, ArrayRange<RenderTreeFram
+        ref var frame = ref frames.Array[position];
+        var result = context.HtmlContentBuilder;
+
+        result.AppendHtml("<");
+        result.AppendHtml(frame.ElementName);
```
Minor optimization - since we know ElementName is `textarea` (and lower/uppercase doesn't matter), this could be reduced to:

```suggestion
        result.AppendHtml("<textarea");
```

SteveSandersonMS commented on src/Mvc/Mvc.ViewFeatures/src/RazorComponents/HtmlRenderer.cs:
```diff
@@ -188,6 +194,34 @@ private int RenderFrames(HtmlRenderingContext context, ArrayRange<RenderTreeFram
+            result.Append(capturedValueAttribute);
+        }
+
+        var afterElement = position + frame.ElementSubtreeLength; // Skip descendants
+        result.AppendHtml("</");
+        result.AppendHtml(frame.ElementName);
+        result.AppendHtml(">");
+        Debug.Assert(afterElement == position + frame.ElementSubtreeLength);
+        return afterElement;
```
Similar to above. Also we don't benefit from the `Debug.Assert` here because we know we just assigned that value, so there's nothing to check really:

```suggestion
        result.AppendHtml("</textarea>");
        return position + frame.ElementSubtreeLength; // Skip descendants
```

SteveSandersonMS commented on src/Mvc/Mvc.ViewFeatures/test/RazorComponents/HtmlRendererTest.cs:
```diff
@@ -316,6 +316,28 @@ public void RenderComponentAsync_MarksSelectedOptionsAsSelected()
+        var serviceProvider = new ServiceCollection().AddSingleton(new RenderFragment(rtb =>
+        {
+            rtb.OpenElement(0, "textarea");
+            rtb.AddAttribute(1, "value", "Hello World!");
```
Would it be possible to change this to something with angle-brackets in just to show the output is correctly encoded? For example:

```suggestion
            rtb.AddAttribute(1, "value", "Hello <html>-encoded content!");
```

... with the assertion being changed to look for `Hello &lt;html&gt;-encoded content!` or similar.

surayya-MS replied:
thanks! done

surayya-MS committed:
Update src/Mvc/Mvc.ViewFeatures/src/RazorComponents/HtmlRenderer.cs

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

SteveSandersonMS reviewed (approved):
Looks like a nice clean solution!

surayya-MS committed:
Update src/Mvc/Mvc.ViewFeatures/src/RazorComponents/HtmlRenderer.cs

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

SteveSandersonMS reviewed (changes requested):
Whoops I just realised there's another important case to consider!

I think this updated logic will now break any code that tries to supply the content as childcontent instead of `value`. Could you add a test case for some logic like `<textarea>Hello</textarea>` to show it renders as `<textarea>Hello</textarea>`? I think the logic as it is in this PR would render as `<textarea></textarea>`, losing the child content.

The rule should be something like "if you supply `value`, that overwrites the child content, but otherwise the child content is left intact."

In the end it might be simpler *not* to have a separate `RenderTextareaElement` method since more of the logic will end up being shared with `RenderElement`, so they might end up being simpler as a single thing.

surayya-MS committed:
changed RendersValueAttributeAsTextContentOfTextareaElement test

surayya-MS committed:
changed RenderElement method and added more tests for textarea

surayya-MS committed:
include comment

surayya-MS commented:
> Whoops I just realised there's another important case to consider!
> 
> I think this updated logic will now break any code that tries to supply the content as childcontent instead of `value`. Could you add a test case for some logic like `<textarea>Hello</textarea>` to show it renders as `<textarea>Hello</textarea>`? I think the logic as it is in this PR would render as `<textarea></textarea>`, losing the child content.
> 
> The rule should be something like "if you supply `value`, that overwrites the child content, but otherwise the child content is left intact."
> 
> In the end it might be simpler _not_ to have a separate `RenderTextareaElement` method since more of the logic will end up being shared with `RenderElement`, so they might end up being simpler as a single thing.

Good catch! I reworked the RenderElement method completely and added more tests to cover all cases.

SteveSandersonMS reviewed (approved):
Nice. Thanks for sorting out all those test cases!
